### ISSUE TITLE
[RMP] Support Offline Batch processing of Inference Pipelines

### ISSUE THREAD

**Issue by jperez999** [NVIDIA-Merlin/Merlin#419]
## Problem:
As a user, I would like to run my merlin systems inference pipeline in an offline setting. This will allow me to conduct rigorous testing and better compare behaviors against other systems, at both operator and system level.  To do this I need to be able to run my merlin systems inference graph without using triton or the configs generated for it. It will require a new operator executor class that runs the ops in python instead of tritonserver. The execution should behave exactly as it does in the tritonserver setting, meaning each operator should be provided same inputs, and return same outputs.

## Goal:
- Run an Inference operator graph without tritonserver.
- Does not require any new user-facing API changes. 
- Execute the same graph, that would be deployed to tritonserver.
- Execute in Python process

## Constraints:
- Use the same merlin systems graph/ops that were created for inference pipeline, that would run on tritonserver
- Swap out the operator executor to python version (non-triton).
- Allow for all types of graphs, supporting multiple chains and parallel running of ALL available operators.

## Starting Point:

- [ ] Create an abstraction that allows for creating a common data type to ensure that different operators with different frameworks can be supported by transforming data -> to common form -> framework specific form. 
- [ ] Create required operator level functions (predict, my_input_type, transform_df, etc...)
                  - [ ] predict -> function that calls predict against model (support any framework)
                  - [ ] my_input_type -> Function that returns the expected type of my inputs. (to be used by operator executor to ensure inputs are in correct format for next operator)
                  - [ ] transform_df -> Function that can be used to house generic operator execution behavior. Should mimic what is happening in inference op logic.
- [ ] Create an executor that will run the ops in the graph.

It is _critical_ that this ticket not only be created, but also kept up to date.  As you work constraints are going to be discovered and should be added to the above list.  Tasks required to complete this project may change.  The goal of the work may even change.  Without a commitment to keeping this ticket up to date the work shouldn't be undertaken.

**Comment by viswa-nvidia** [comment]
Assignees will be Karl / Adam.

**Comment by sohn21c** [comment]
This is a prerequisite for cross-FW evaluation

**Comment by nv-alaiacano** [comment]
My impression is that batch inference for _models_ is required for cross-FW evaluation, not the full batch inference for a _system_. The additional steps in the Systems' computation graph (QueryFeast, QueryFaiss, Softmax, filtering, etc) would likely not be required for batch inference on a single Model. Batch inference for the model would have a simpler "training data in -> predictions out" process, which would likely be a step in the Systems graph.

Perhaps we should first build the batch inference functionality (apply nvt transform + use model to predict) including the output format schema, and then that functionality could be shared in cross-FW evaluation and systems-wide batch prediction.

**Comment by karlhigley** [comment]
We do have some batch prediction functionality for models already, but it's not quite structured in a way that would make it a reasonable foundation for batch processing of graphs. I think we could massage it in that direction though and try to standardize how batch graph processing works in Merlin Core by taking what exists and refactoring it in the right direction.

**Comment by bschifferer** [comment]
@karlhigley do you think we should add an example for it?

**Comment by karlhigley** [comment]
I think we should add an example for every new piece of significant functionality (i.e. almost all roadmap issues.)

**Comment by jperez999** [comment]
https://github.com/NVIDIA-Merlin/core/pull/352
https://github.com/NVIDIA-Merlin/systems/pull/376

**Comment by karlhigley** [comment]
@jperez999 Could you add appropriate tasks to the list in the description?

**Comment by jperez999** [comment]
Need to be able to swap out certain operators, based on runtime. I.e. when running daskexecutor for offline batch it is not necessary to run the feature store operator unless we are testing against it. You could run a dataset merge operator instead using offline features stored in a parquet file.  Please refer to task list created for further tracking
